File: clients/ios/Lextures/Core/LMS/CourseArchivedContentLogic.swift
import Foundation

/// Archived module item list + restore helpers (M13.12).
enum CourseArchivedContentLogic {
    struct ArchivedContentRow: Identifiable, Hashable, Equatable {
        var id: String
        var title: String
        var kind: String
        var kindLabelKey: String
        var moduleTitle: String
        var archivedAt: String?
    }

    private static let restorableKinds: Set<String> = [
        "heading",
        "content_page",
        "assignment",
        "quiz",
        "external_link",
    ]

    static func canViewArchivedContent(courseCode: String, permissions: [String]) -> Bool {
        CourseSettingsLogic.canManageCourse(courseCode: courseCode, permissions: permissions)
    }

    static func cacheKeyArchivedStructure(courseCode: String) -> String {
        "course:\(courseCode):archived-structure"
    }

    static func moduleTitleById(from items: [CourseStructureItem]) -> [String: String] {
        var titles: [String: String] = [:]
        for item in items where item.kind == "module" && item.parentId == nil {
            titles[item.id] = item.title
        }
        return titles
    }

    static func archivedRows(from items: [CourseStructureItem]) -> [ArchivedContentRow] {
        let modules = moduleTitleById(from: items)
        return items
            .filter { item in
                item.archived == true
                    && item.parentId != nil
                    && restorableKinds.contains(item.kind)
            }
            .sorted { lhs, rhs in
                if lhs.sortOrder != rhs.sortOrder {
                    return lhs.sortOrder < rhs.sortOrder
                }
                return lhs.title.localizedCaseInsensitiveCompare(rhs.title) == .orderedAscending
            }
            .map { item in
                ArchivedContentRow(
                    id: item.id,
                    title: item.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                        ? L.text("mobile.emDash")
                        : item.title,
                    kind: item.kind,
                    kindLabelKey: kindLabelKey(for: item.kind),
                    moduleTitle: item.parentId.flatMap { modules[$0] } ?? L.text("mobile.emDash"),
                    archivedAt: item.updatedAt
                )
            }
    }

    static func itemsAfterRestore(items: [CourseStructureItem], removedId: String) -> [CourseStructureItem] {
        items.filter { $0.id != removedId }
    }

    static func kindLabelKey(for kind: String) -> String {
        switch kind {
        case "heading":
            return "mobile.courseSettings.archivedContent.kind.heading"
        case "content_page":
            return "mobile.courseSettings.archivedContent.kind.contentPage"
        case "assignment":
            return "mobile.courseSettings.archivedContent.kind.assignment"
        case "quiz":
            return "mobile.courseSettings.archivedContent.kind.quiz"
        case "external_link":
            return "mobile.courseSettings.archivedContent.kind.externalLink"
        default:
            return "mobile.courseSettings.archivedContent.kind.other"
        }
    }

    static func formatArchivedAt(_ iso: String?) -> String {
        let formatted = DateFormatting.formatDateTime(iso)
        return formatted.isEmpty ? L.text("mobile.emDash") : formatted
    }

    static func userFacingError(_ error: Error) -> String {
        if let apiError = error as? APIError {
            return apiError.errorDescription ?? L.text("mobile.courseSettings.archivedContent.genericError")
        }
        return error.localizedDescription
    }
}

struct CourseStructureItemPatch: Encodable {
    var archived: Bool?
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseArchivedContentLogic.kt
package com.lextures.android.core.lms

/** Archived module item list + restore helpers (M13.12). */
object CourseArchivedContentLogic {
    private val restorableKinds = setOf(
        "heading",
        "content_page",
        "assignment",
        "quiz",
        "external_link",
    )

    data class ArchivedContentRow(
        val id: String,
        val title: String,
        val kind: String,
        val kindLabelKey: String,
        val moduleTitle: String,
        val archivedAt: String?,
    )

    fun canViewArchivedContent(courseCode: String, permissions: List<String>): Boolean =
        CourseSettingsLogic.canManageCourse(courseCode, permissions)

    fun cacheKeyArchivedStructure(courseCode: String): String = "course:$courseCode:archived-structure"

    fun moduleTitleById(items: List<CourseStructureItem>): Map<String, String> =
        items
            .asSequence()
            .filter { it.kind == "module" && it.parentId == null }
            .associate { it.id to it.title }

    fun archivedRows(items: List<CourseStructureItem>): List<ArchivedContentRow> {
        val modules = moduleTitleById(items)
        return items
            .asSequence()
            .filter { item ->
                item.archived == true &&
                    item.parentId != null &&
                    item.kind in restorableKinds
            }
            .sortedWith(
                compareBy<CourseStructureItem> { it.sortOrder }
                    .thenBy(String.CASE_INSENSITIVE_ORDER) { it.title },
            )
            .map { item ->
                ArchivedContentRow(
                    id = item.id,
                    title = item.title.trim().ifEmpty { "—" },
                    kind = item.kind,
                    kindLabelKey = kindLabelKey(item.kind),
                    moduleTitle = item.parentId?.let(modules::get) ?: "—",
                    archivedAt = item.updatedAt,
                )
            }
            .toList()
    }

    fun itemsAfterRestore(items: List<CourseStructureItem>, removedId: String): List<CourseStructureItem> =
        items.filter { it.id != removedId }

    fun kindLabelKey(kind: String): String = when (kind) {
        "heading" -> "mobile.courseSettings.archivedContent.kind.heading"
        "content_page" -> "mobile.courseSettings.archivedContent.kind.contentPage"
        "assignment" -> "mobile.courseSettings.archivedContent.kind.assignment"
        "quiz" -> "mobile.courseSettings.archivedContent.kind.quiz"
        "external_link" -> "mobile.courseSettings.archivedContent.kind.externalLink"
        else -> "mobile.courseSettings.archivedContent.kind.other"
    }

    fun formatArchivedAt(raw: String?): String {
        val formatted = LmsDates.shortDateTime(raw)
        return formatted.ifEmpty { "—" }
    }

    fun userFacingError(error: Throwable): String = error.message ?: "generic"
}
